### PR TITLE
[GH-546] Add CoM User CRUD

### LINKED ISSUE #546: Add Curse of Mirra user crud
We need to:

- Create a Curse of Mirra User.
- Add username and profile_picture fields.
- Associate it with the GoogleUser schema.
- Add CRUD operations.

### PR DESCRIPTION
**Pull request by Nico-Sanchez**
## Motivation
Closes #546 
We need to allow users to change their information (username, profile pic).
This User must be associated with the one created with google.

## Changes
- [Add google user assoc and profile picture field to users table](https://github.com/lambdaclass/mirra_backend/pull/548/commits/9274011e62da5120692ad485d4a88c4bcaf2c0d7)
- [Add update_user/2 function to update a user's params](https://github.com/lambdaclass/mirra_backend/pull/548/commits/e4b297f92ceaf0d8768bd4b6d970c77475546de9)
- [Add new fields and assocs to User and GoogleUser schemas](https://github.com/lambdaclass/mirra_backend/pull/548/commits/0fae34239872d8c4f1562ace53a8c4fbca928148)
- [Add new endpoints and controller to handle username and profile picture](https://github.com/lambdaclass/mirra_backend/pull/548/commits/0ec87b02793cff03ad3f67b0128189b19c17f463)
- [Fix google_user and user associations and changeset rules](https://github.com/lambdaclass/mirra_backend/pull/548/commits/027eea4abf8896f02c8ad74a4712fde195758ac8)
- [Create User when GoogleUser is created](https://github.com/lambdaclass/mirra_backend/pull/548/commits/14701eddc26e4111559c1edfbe113b5671b143c4)

### CODE REVIEW

**Comment by AminArria** [comment]
Do we have a business decision about usernames and profile pictures across games? Are they supposed to be per game or shared?

File: apps/game_backend/lib/game_backend/utils.ex [hunk 1]
```diff
@@ -0,0 +1,8 @@
+defmodule GameBackend.Utils do
+  @moduledoc """
+  Helper module
+  """
+
+  def get_game_id("curse_of_mirra"), do: 1
+  def get_game_id("champions_of_mirra"), do: 2
```
Comment (AminArria) [inline apps/game_backend/lib/game_backend/utils.ex:7]: Use atoms, not strings

File: apps/game_backend/priv/repo/migrations/20240429180719_add_google_user_assoc_to_users.exs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+defmodule GameBackend.Repo.Migrations.AddGoogleUserAssocToUsers do
+  use Ecto.Migration
+
+  def change do
+    alter table(:users) do
+      add :google_user_id, references(:google_users, on_delete: :delete_all)
+      add :profile_picture, :string
```
Comment (AminArria) [inline apps/game_backend/priv/repo/migrations/20240429180719_add_google_user_assoc_to_users.exs:7]: What is this supposed to store?

File: apps/game_backend/lib/game_backend/utils.ex (hunk 1 above)
Comment (AminArria) [inline apps/game_backend/lib/game_backend/utils.ex:7]: Use atoms, not strings

File: apps/gateway/lib/gateway/controllers/user_controller.ex [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  end
+
+  defp update_user_params(conn, user_id, params) do
+    conn = put_resp_content_type(conn, "application/json")
+
+    with {:ok, user} <- Users.get_user(user_id),
+         {:ok, user} <- Users.update_user(user, params) do
+      send_resp(conn, 200, Jason.encode!(user))
+    else
+      {:error, :not_found} -> send_resp(conn, 200, Jason.encode!(%{"error" => "not found"}))
```
Comment (AminArria) [inline apps/gateway/lib/gateway/controllers/user_controller.ex:23]: This should be a 404

File: apps/gateway/lib/gateway/controllers/user_controller.ex [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+      {:error, _changeset} -> send_resp(conn, 200, Jason.encode!(%{"error" => "failed to update"}))
```
Comment (AminArria) [inline apps/gateway/lib/gateway/controllers/user_controller.ex:24]: This should be a 400

File: apps/gateway/lib/gateway/router.ex [hunk 5]
```diff
@@ -15,6 +15,13 @@ defmodule Gateway.Router do
     get "/:provider/token/:token_id", Controllers.AuthController, :validate_token
   end
 
+  scope "/users", Gateway do
+    pipe_through :api
+
+    post "/users/:user_id/change_username/:username", Controllers.UserController, :change_username
+    post "/users/:user_id/change_profile_picture/:profile_picture", Controllers.UserController, :change_profile_picture
```
Comment (AminArria) [inline apps/gateway/lib/gateway/router.ex:22]: Both endpoints should be either PUT or PATCH

Also, both of this endpoints could be the same `/users/:user_id` and `update_user_params` in the controller, the clients would send the field to modify in the body of the request

File: apps/gateway/lib/gateway/controllers/user_controller.ex [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  def change_username(conn, params) do
+    update_user_params(conn, params.user_id, %{"username" => params.username})
+  end
+
+  def change_profile_picture(conn, params) do
+    update_user_params(conn, params.user_id, %{"profile_picture" => params.profile_picture})
+  end
+
+  defp update_user_params(conn, user_id, params) do
+    conn = put_resp_content_type(conn, "application/json")
```
Comment (AminArria) [inline apps/gateway/lib/gateway/controllers/user_controller.ex:17]: Needed?

**Review by AminArria — DISMISSED** [review]
Do we have a business decision about usernames and profile pictures across games? Are they supposed to be per game or shared?

File: apps/game_backend/priv/repo/migrations/20240429180719_add_google_user_assoc_to_users.exs (hunk 2 above)
Comment (Nico-Sanchez) [reply apps/game_backend/priv/repo/migrations/20240429180719_add_google_user_assoc_to_users.exs:7]: Just a string. i.e.: "picture_1", "picture_2", etc. It's like characters.
At least for now.

File: apps/game_backend/lib/game_backend/utils.ex (hunk 1 above)
Comment (Nico-Sanchez) [reply apps/game_backend/lib/game_backend/utils.ex:7]: Alright. Could I ask why?

File: apps/gateway/lib/gateway/controllers/user_controller.ex (hunk 3 above)
Comment (Nico-Sanchez) [reply apps/gateway/lib/gateway/controllers/user_controller.ex:23]: Lol you're right. Changed!

File: apps/gateway/lib/gateway/controllers/user_controller.ex (hunk 4 above)
Comment (Nico-Sanchez) [reply apps/gateway/lib/gateway/controllers/user_controller.ex:24]: Done!

File: apps/gateway/lib/gateway/router.ex (hunk 5 above)
Comment (Nico-Sanchez) [reply apps/gateway/lib/gateway/router.ex:22]: Sure! Done!

File: apps/gateway/lib/gateway/controllers/user_controller.ex (hunk 6 above)
Comment (Nico-Sanchez) [reply apps/gateway/lib/gateway/controllers/user_controller.ex]: No. Deleted.

File: apps/game_backend/lib/game_backend/utils.ex (hunk 1 above)
[the remark this answers is not in the document]
Comment (Nico-Sanchez) [reply apps/game_backend/lib/game_backend/utils.ex]: Nice one. Done!

**Review by agustinesco — DISMISSED** [review]
LGTM!

File: apps/game_backend/lib/game_backend/users.ex [hunk 7]
```diff
@@ -38,22 +39,40 @@ defmodule GameBackend.Users do
...
   end
 
+  @doc """
+  Updates the given user by given params.
+  Returns {:ok, User}.
+  Returns {:error, Changeset} if update transaction fails.
+
+  ## Examples
+
+      iex> update_user("51646f3a-d9e9-4ce6-8341-c90b8cad3bdf")
```
Comment (lotuuu) [inline apps/game_backend/lib/game_backend/users.ex:65]: This is missing a parameter!

**Review by lotuuu — CHANGES_REQUESTED** [review]

**Review by lotuuu — APPROVED** [review]

File: apps/game_backend/lib/game_backend/users.ex (hunk 7 above)
Comment (agustinesco) [reply apps/game_backend/lib/game_backend/users.ex]: fixed!

**Review by agustinesco — APPROVED** [review]
LGTM!

**Pull request merged**